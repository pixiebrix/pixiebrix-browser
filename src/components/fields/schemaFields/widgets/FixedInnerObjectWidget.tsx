/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { useMemo } from "react";
import { type SchemaFieldProps } from "@/components/fields/schemaFields/propTypes";
import genericOptionsFactory from "@/components/fields/schemaFields/genericOptionsFactory";
import { type Schema } from "@/core";
import WorkshopMessageWidget from "@/components/fields/schemaFields/widgets/WorkshopMessageWidget";

/**
 * A widget for rendering a nested object with fixed properties.
 * @param props
 * @constructor
 * @see isCustomizableObjectSchema
 */
const FixedInnerObjectWidget: React.FC<SchemaFieldProps> = (props) => {
  const { name, schema } = props;

  const Fields = useMemo(() => {
    let objectSchema = schema;

    if (schema.oneOf) {
      const matches = objectSchema.oneOf.filter(
        (x) => typeof x !== "boolean" && x.type === "object"
      );

      if (matches.length > 1) {
        return WorkshopMessageWidget;
      }

      objectSchema = matches[0] as Schema;
    }

    if (schema.allOf || schema.anyOf) {
      return WorkshopMessageWidget;
    }

    if (objectSchema.type !== "object") {
      // Bail if somehow a non-object schema was passed in
      return WorkshopMessageWidget;
    }

    return genericOptionsFactory(objectSchema, null, {
      preserveSchemaOrder: true,
    });
  }, [schema]);

  return <Fields name={name} />;
};

export default FixedInnerObjectWidget;
