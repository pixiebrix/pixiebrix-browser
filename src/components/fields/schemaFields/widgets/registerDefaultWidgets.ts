/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import UrlMatchPatternWidget from "./UrlMatchPatternWidget";
import UrlPatternWidget from "@/pageEditor/components/UrlPatternWidget";
import ArrayWidget from "./ArrayWidget";
import BooleanWidget from "./BooleanWidget";
import IntegerWidget from "./IntegerWidget";
import NumberWidget from "./NumberWidget";
import ObjectWidget from "./ObjectWidget";
import OmitFieldWidget from "./OmitFieldWidget";
import SchemaSelectWidget from "./SchemaSelectWidget";
import TemplateToggleWidget from "./TemplateToggleWidget";
import TextWidget from "./TextWidget";
import UnsupportedWidget from "./UnsupportedWidget";
import widgetsRegistry from "./widgetsRegistry";
import WorkshopMessageWidget from "./WorkshopMessageWidget";
import SelectorMatchWidget from "@/pageEditor/components/SelectorMatchWidget";
import CssClassWidget from "./CssClassWidget";
import DatabaseWidget from "./DatabaseWidget";
import PasswordWidget from "./PasswordWidget";
import IconWidget from "@/components/fields/IconWidget";

const defaultWidgets = {
  ArrayWidget,
  BooleanWidget,
  IntegerWidget,
  NumberWidget,
  ObjectWidget,
  OmitFieldWidget,
  SchemaSelectWidget,
  TemplateToggleWidget,
  TextWidget,
  UnsupportedWidget,
  UrlMatchPatternWidget,
  UrlPatternWidget,
  SelectorMatchWidget,
  CssClassWidget,
  WorkshopMessageWidget,
  DatabaseWidget,
  IconWidget,
  PasswordWidget,
} as const;

function registerDefaultWidgets() {
  Object.assign(widgetsRegistry, defaultWidgets);
  for (const [name, widget] of Object.entries(defaultWidgets)) {
    if (!widget) {
      throw new Error(
        `Error registering default widgets. ${name} is undefined. Is there a circular dependency?`
      );
    }
  }
}

export default registerDefaultWidgets;
