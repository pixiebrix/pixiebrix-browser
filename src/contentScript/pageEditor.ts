/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { serializeError } from "serialize-error";
import {
  blockReducer,
  type IntermediateState,
  type ReduceOptions,
} from "@/runtime/reducePipeline";
import { type BlockConfig } from "@/blocks/types";
import { cloneDeep } from "lodash";
import ConsoleLogger from "@/utils/ConsoleLogger";
import { type SerializableResponse } from "@/pageScript/messenger/pigeon";
import apiVersionOptions from "@/runtime/apiVersionOptions";
import { $safeFind } from "@/helpers";
import { clearDynamicElements } from "@/contentScript/pageEditor/dynamic";
import { reactivateTab } from "./lifecycle";
import {
  BusinessError,
  CancelError,
  NoRendererError,
} from "@/errors/businessErrors";
import { uuidv4 } from "@/types/helpers";
import { type PanelPayload } from "@/sidebar/types";
import { HeadlessModeError } from "@/blocks/errors";
import { showTemporarySidebarPanel } from "@/contentScript/sidebarController";
import { stopInspectingNativeHandler } from "./pageEditor/elementPicker";
import { showModal } from "@/blocks/transformers/ephemeralForm/modalUtils";
import { createFrameSource } from "@/blocks/transformers/temporaryInfo/DisplayTemporaryInfo";
import { waitForTemporaryPanel } from "@/blocks/transformers/temporaryInfo/temporaryPanelProtocol";
import { type ApiVersion, type BlockArgsContext } from "@/types/runtimeTypes";
import { type UUID } from "@/types/stringTypes";
import { type RegistryId } from "@/types/registryTypes";

export type RunBlockArgs = {
  /**
   * The runtime API version to use
   */
  apiVersion: ApiVersion;
  /**
   * The IBlock configuration.
   */
  blockConfig: BlockConfig;
  /**
   * Context to render the BlockArg, should include @input, @options, and service context
   * @see makeServiceContext
   */
  context: BlockArgsContext;
  /**
   * Root jQuery selector to determine the root if the rootMode is "inherit".
   * @see BlockConfig.rootMode
   */
  rootSelector: string | undefined;
};

/**
 * Run a single block (e.g., for generating output previews)
 * @see BlockPreview
 */
export async function runBlock({
  blockConfig,
  context,
  apiVersion,
  rootSelector,
}: RunBlockArgs): Promise<unknown> {
  const versionOptions = apiVersionOptions(apiVersion);

  if (!versionOptions.explicitDataFlow) {
    throw new BusinessError(
      "Preview only supported for mods using runtime v2 or later"
    );
  }

  const state: IntermediateState = {
    context,
    // Can pick any index. It's only used for adding context to log messages, and we're disabling value logging
    // below with `logValues: false`
    index: 0,
    // Force isLastBlock so blockReducer does not complain about the outputKey being forced to undefined
    isLastBlock: true,
    // `root` is over-ridden below if rootSelector is provided
    root: document,
    // We're forcing apiVersion: 2 or higher above values must come from the context
    previousOutput: {},
  };

  if (rootSelector) {
    // Handle non-document contexts. If the selector is unique, this gives the root that would be available at runtime.
    // Differences in behavior:
    // - For triggers, the PixieBrix looks for the closest ancestor to the DOM event target matching the selector
    //   See TriggerExtensionPoint.eventHandler for reference
    // - For multi-menus (not currently available in the Page Editor), the below logic returns an arbitrary menu
    const rootElement = $safeFind(rootSelector);
    if (rootElement.length > 0) {
      state.root = rootElement.get(0);
    }
  }

  const options: ReduceOptions = {
    ...versionOptions,
    branches: [],
    headless: true,
    logValues: false,
    logger: new ConsoleLogger(),
    // Excluding runId will prevent the run from being stored in traces
    runId: null,
    extensionId: null,
  };

  // Exclude the outputKey so that `output` is the output of the brick. Alternatively we could have taken then
  // value from the context[outputKey] from the return value of blockReducer
  const { output } = await blockReducer(
    { ...blockConfig, outputKey: undefined },
    state,
    options
  );

  return cloneDeep(output) as SerializableResponse;
}

type Location = "modal" | "panel";

/**
 * Run a single renderer (e.g. - for running a block preview)
 *
 * Renderers need to be run with try-catch, catch the HeadlessModeError, and
 * use that to send the panel payload to the sidebar (or other target)
 * @see SidebarExtensionPoint
 *  starting on line 184, the call to reduceExtensionPipeline(),
 *  wrapped in a try-catch
 * @see executeBlockWithValidatedProps
 *  starting on line 323, the runRendererPipeline() function
 *
 * Note: Currently only implemented for the temporary sidebar panels
 * @see useDocumentPreviewRunBlock
 */
export async function runRendererBlock({
  extensionId,
  blueprintId,
  runId,
  title,
  args,
  location,
}: {
  extensionId: UUID;
  blueprintId: RegistryId | null;
  runId: UUID;
  title: string;
  args: RunBlockArgs;
  location: Location;
}): Promise<void> {
  const nonce = uuidv4();

  let payload: PanelPayload;
  try {
    await runBlock(args);
    // We're expecting a HeadlessModeError (or other error) to be thrown in the line above
    // noinspection ExceptionCaughtLocallyJS
    throw new NoRendererError();
  } catch (error) {
    if (error instanceof HeadlessModeError) {
      payload = {
        key: nonce,
        blockId: error.blockId,
        args: error.args,
        ctxt: error.ctxt,
        extensionId,
        runId,
      };
    } else {
      payload = {
        key: nonce,
        error: serializeError(error),
        extensionId,
        runId,
      };
    }

    if (location === "panel") {
      showTemporarySidebarPanel({
        // Pass extension id so previous run is cancelled
        extensionId,
        blueprintId,
        nonce,
        heading: title,
        payload,
      });
    } else if (location === "modal") {
      const controller = new AbortController();
      const url = await createFrameSource(nonce, "modal");

      showModal({ url, controller });

      try {
        await waitForTemporaryPanel(nonce, {
          extensionId,
          blueprintId,
          nonce,
          heading: title,
          payload,
        });
      } catch (error) {
        // Match behavior of Display Temporary Info
        if (error instanceof CancelError) {
          // NOP
        } else {
          throw error;
        }
      } finally {
        controller.abort();
      }
    } else {
      // eslint-disable-next-line @typescript-eslint/restrict-template-expressions -- dynamic check for never
      throw new Error(`Support for previewing in ${location} not implemented`);
    }
  }
}

export async function resetTab(): Promise<void> {
  stopInspectingNativeHandler();
  await clearDynamicElements({});
  await reactivateTab();
}
