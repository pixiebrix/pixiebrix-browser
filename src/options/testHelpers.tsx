/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { configureStore } from "@reduxjs/toolkit";
import { authSlice } from "@/auth/authSlice";
import extensionsSlice from "@/store/extensionsSlice";
import settingsSlice from "@/store/settingsSlice";
import { blueprintModalsSlice } from "@/options/pages/blueprints/modals/blueprintModalsSlice";
import { createRenderWithWrappers } from "@/testUtils/testHelpers";
import blueprintsSlice from "@/options/pages/blueprints/blueprintsSlice";
import { recipesSlice } from "@/recipes/recipesSlice";

const renderWithWrappers = createRenderWithWrappers(() =>
  configureStore({
    reducer: {
      auth: authSlice.reducer,
      settings: settingsSlice.reducer,
      options: extensionsSlice.reducer,
      blueprintModals: blueprintModalsSlice.reducer,
      blueprints: blueprintsSlice.reducer,
      recipes: recipesSlice.reducer,
    },
  })
);

// eslint-disable-next-line import/export -- re-export RTL
export * from "@testing-library/react";
// eslint-disable-next-line import/export -- override render
export { renderWithWrappers as render };
