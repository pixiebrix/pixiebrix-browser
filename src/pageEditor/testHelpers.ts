/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { configureStore } from "@reduxjs/toolkit";
import { authSlice } from "@/auth/authSlice";
import extensionsSlice from "@/store/extensionsSlice";
import servicesSlice from "@/store/servicesSlice";
import settingsSlice from "@/store/settingsSlice";
import { editorSlice } from "@/pageEditor/slices/editorSlice";
import sessionSlice from "@/pageEditor/slices/sessionSlice";
import { savingExtensionSlice } from "@/pageEditor/panes/save/savingExtensionSlice";
import runtimeSlice from "@/pageEditor/slices/runtimeSlice";
import { logSlice } from "@/components/logViewer/logSlice";
import {
  createRenderHookWithWrappers,
  createRenderWithWrappers,
} from "@/testUtils/testHelpers";
import analysisSlice from "@/analysis/analysisSlice";
import pageEditorAnalysisManager from "./analysisManager";
import { tabStateSlice } from "@/pageEditor/tabState/tabStateSlice";
import { appApi } from "@/services/api";
import { modDefinitionsSlice } from "@/modDefinitions/modDefinitionsSlice";
import { modDefinitionsMiddleware } from "@/modDefinitions/modDefinitionsListenerMiddleware";

const configureStoreForTests = () =>
  configureStore({
    reducer: {
      auth: authSlice.reducer,
      options: extensionsSlice.reducer,
      services: servicesSlice.reducer,
      settings: settingsSlice.reducer,
      editor: editorSlice.reducer,
      session: sessionSlice.reducer,
      savingExtension: savingExtensionSlice.reducer,
      runtime: runtimeSlice.reducer,
      logs: logSlice.reducer,
      analysis: analysisSlice.reducer,
      tabState: tabStateSlice.reducer,
      modDefinitions: modDefinitionsSlice.reducer,
      [appApi.reducerPath]: appApi.reducer,
    },
    middleware(getDefaultMiddleware) {
      /* eslint-disable unicorn/prefer-spread -- It's not Array#concat, can't use spread */
      return getDefaultMiddleware()
        .concat(appApi.middleware)
        .concat(pageEditorAnalysisManager.middleware)
        .concat(modDefinitionsMiddleware);
      /* eslint-enable unicorn/prefer-spread */
    },
  });

const renderWithWrappers = createRenderWithWrappers(configureStoreForTests);
const renderHookWithWrappers = createRenderHookWithWrappers(
  configureStoreForTests
);

// eslint-disable-next-line import/export -- re-export RTL
export * from "@testing-library/react";
// eslint-disable-next-line import/export -- override render
export { renderWithWrappers as render };
export { renderHookWithWrappers as renderHook };
