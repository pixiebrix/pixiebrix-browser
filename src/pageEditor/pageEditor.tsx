/*
 * Copyright (C) 2024 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import "@/vendors/bootstrapWithoutRem.css";
import "@/utils/global.scss";
import "@/utils/layout.scss";

import "@/extensionContext";
import "@/development/darkMode.js";

import { initMessengerLogging } from "@/development/messengerLogging";

import ReactDOM from "react-dom";
import React from "react";
import Panel from "@/pageEditor/Panel";
import { watchNavigation } from "@/pageEditor/protocol";
import { initToaster } from "@/utils/notify";
import { initRuntimeLogging } from "@/development/runtimeLogging";
import { initPerformanceMonitoring } from "@/telemetry/performance";
import { setPlatform } from "@/platform/platformContext";
import extensionPagePlatform from "@/extensionPages/extensionPagePlatform";

async function init() {
  setPlatform(extensionPagePlatform);
  void initMessengerLogging();
  void initRuntimeLogging();
  try {
    await initPerformanceMonitoring();
  } catch (error) {
    console.error("Failed to initialize performance monitoring", error);
  }

  watchNavigation();
  initToaster();

  ReactDOM.render(<Panel />, document.querySelector("#container"));
}

void init();
