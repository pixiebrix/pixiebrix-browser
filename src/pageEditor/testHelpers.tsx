/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { configureStore } from "@reduxjs/toolkit";
import { authSlice } from "@/auth/authSlice";
import extensionsSlice from "@/store/extensionsSlice";
import servicesSlice from "@/store/servicesSlice";
import settingsSlice from "@/store/settingsSlice";
import { editorSlice } from "@/pageEditor/slices/editorSlice";
import sessionSlice from "@/pageEditor/slices/sessionSlice";
import { savingExtensionSlice } from "@/pageEditor/panes/save/savingExtensionSlice";
import runtimeSlice from "@/pageEditor/slices/runtimeSlice";
import { logSlice } from "@/components/logViewer/logSlice";
import { createRenderWithWrappers } from "@/testUtils/testHelpers";
import { appApi } from "@/services/api";

const renderWithWrappers = createRenderWithWrappers(() =>
  configureStore({
    reducer: {
      auth: authSlice.reducer,
      options: extensionsSlice.reducer,
      services: servicesSlice.reducer,
      settings: settingsSlice.reducer,
      editor: editorSlice.reducer,
      session: sessionSlice.reducer,
      savingExtension: savingExtensionSlice.reducer,
      runtime: runtimeSlice.reducer,
      logs: logSlice.reducer,
      [appApi.reducerPath]: appApi.reducer,
    },
  })
);

// eslint-disable-next-line import/export -- re-export RTL
export * from "@testing-library/react";
// eslint-disable-next-line import/export -- override render
export { renderWithWrappers as render };
