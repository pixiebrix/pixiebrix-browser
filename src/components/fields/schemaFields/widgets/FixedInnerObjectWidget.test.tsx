/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { render } from "@/pageEditor/testHelpers";
import React from "react";
import FixedInnerObjectWidget from "@/components/fields/schemaFields/widgets/FixedInnerObjectWidget";
import { type Schema } from "@/core";
import registerDefaultWidgets from "@/components/fields/schemaFields/widgets/registerDefaultWidgets";

beforeAll(() => {
  registerDefaultWidgets();
});

describe("FixedInnerObjectWidget", () => {
  it("bail on allOf", () => {
    const wrapper = render(
      <FixedInnerObjectWidget name="test" schema={{ allOf: [] }} />
    );
    expect(wrapper.queryByDisplayValue("Use Workshop to edit")).not.toBeNull();
  });

  it("handle simple object", () => {
    const schema = {
      type: "object",
      properties: { foo: { type: "string" } },
    } as Schema;
    const wrapper = render(
      <FixedInnerObjectWidget name="test" schema={schema} />,
      {
        initialValues: { test: { foo: "bar" } },
      }
    );
    expect(wrapper.queryByDisplayValue("Use Workshop to edit")).toBeNull();
    expect(wrapper.container).toMatchSnapshot();
  });

  it("handle oneOf object", () => {
    const schema = {
      oneOf: [
        { type: "object", properties: { foo: { type: "string" } } },
        { type: "string" },
      ],
    } as Schema;
    const wrapper = render(
      <FixedInnerObjectWidget name="test" schema={schema} />,
      {
        initialValues: { test: { foo: "bar" } },
      }
    );
    expect(wrapper.queryByDisplayValue("Use Workshop to edit")).toBeNull();
    expect(wrapper.container).toMatchSnapshot();
  });
});
