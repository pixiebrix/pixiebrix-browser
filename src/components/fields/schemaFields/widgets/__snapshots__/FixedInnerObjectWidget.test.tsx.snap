// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FixedInnerObjectWidget handle oneOf object 1`] = `
<div>
  <form
    action="#"
  >
    <div
      class="formGroup form-group row"
    >
      <label
        class="label form-label col-form-label col-xl-2 col-lg-3"
        for="test.foo"
      >
        foo
      </label>
      <div
        class="col-xl-10 col-lg-9"
      >
        <div
          class="root"
        >
          <div
            class="field"
          >
            <textarea
              class="form-control"
              id="test.foo"
              name="test.foo"
              rows="1"
            >
              bar
            </textarea>
          </div>
          <div
            class="dropdown dropdown"
            data-test-selected="Text"
            data-testid="toggle-test.foo"
          >
            <button
              aria-expanded="false"
              aria-haspopup="true"
              class="dropdown-toggle btn btn-link"
              type="button"
            >
              <span
                class="symbol"
              >
                <test-file-stub
                  classname="root"
                />
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`FixedInnerObjectWidget handle simple object 1`] = `
<div>
  <form
    action="#"
  >
    <div
      class="formGroup form-group row"
    >
      <label
        class="label form-label col-form-label col-xl-2 col-lg-3"
        for="test.foo"
      >
        foo
      </label>
      <div
        class="col-xl-10 col-lg-9"
      >
        <div
          class="root"
        >
          <div
            class="field"
          >
            <textarea
              class="form-control"
              id="test.foo"
              name="test.foo"
              rows="1"
            >
              bar
            </textarea>
          </div>
          <div
            class="dropdown dropdown"
            data-test-selected="Text"
            data-testid="toggle-test.foo"
          >
            <button
              aria-expanded="false"
              aria-haspopup="true"
              class="dropdown-toggle btn btn-link"
              type="button"
            >
              <span
                class="symbol"
              >
                <test-file-stub
                  classname="root"
                />
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
